components/function-controller/internal/controllers/serverless/horizontalpodautoscaler.go

```diff
@@ -0,0 +1,108 @@
+package serverless
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	serverlessv1alpha1 "github.com/kyma-project/kyma/components/function-controller/pkg/apis/serverless/v1alpha1"
+)
+
+func (r *FunctionReconciler) isOnHorizontalPodAutoscalerChange(instance *serverlessv1alpha1.Function, hpas []autoscalingv1.HorizontalPodAutoscaler, deployments []appsv1.Deployment) bool {
+	if len(deployments) == 0 {
+		return false
+	}
+
+	newHpa := r.buildHorizontalPodAutoscaler(instance, deployments[0].GetName())
+	return !(len(hpas) == 1 &&
+		r.equalHorizontalPodAutoscalers(hpas[0], newHpa))
+}
+
+func (r *FunctionReconciler) onHorizontalPodAutoscalerChange(ctx context.Context, log logr.Logger, instance *serverlessv1alpha1.Function, hpas []autoscalingv1.HorizontalPodAutoscaler, deploymentName string) (ctrl.Result, error) {
+	newHpa := r.buildHorizontalPodAutoscaler(instance, deploymentName)
+
+	switch {
+	case len(hpas) == 0:
+		return r.createHorizontalPodAutoscaler(ctx, log, instance, newHpa)
+	case len(hpas) > 1:
+		log.Info("Too many HorizontalPodAutoscalers for function")
+		return r.updateStatus(ctx, ctrl.Result{RequeueAfter: r.config.RequeueDuration}, instance, serverlessv1alpha1.Condition{
+			Type:               serverlessv1alpha1.ConditionRunning,
+			Status:             corev1.ConditionFalse,
+			LastTransitionTime: metav1.Now(),
+			Reason:             serverlessv1alpha1.ConditionReasonHorizontalPodAutoscalerError,
+			Message:            fmt.Sprintf("HorizontalPodAutoscaler %s failed", instance.GetName()),
```

Comment:
Please add also the message on why it failed. Otherwise, you will need to look into the logs for a reason...